### PR TITLE
Mount remote storage with FUSE

### PR DESCRIPTION
**Pull request by PhilippeMoussalli**
This PR replaces the current way of interacting with cloud storage. It uses FUSE to mount the remote cloud storage as a local file system. This enables us to read/write files directly from files stored remotely instead of downloading them to storage. 

Check out this [document](https://docs.google.com/document/d/1Yn8ayLBovv2PHlLQhsxp81Wzp8I23pOjW4xlMkqwmjs/edit ) for more description on FUSE and considered alternatives 

ToDos: 

- [X ] Add a Dockerfile for installation of fuse 
- [ X] Read/write files directly from the mount direction (right now they are being written to a local directory and copied over to the mount location) 
- [ X] Add necessary changes to the HF dataset component 
- [ ] Migrate changes to existing pipeline 

Streaming functionality will be implemented in a separate PR

### CODE REVIEW

File: docker/gcsfuse/Dockerfile
Comment (RobbeSneyders) [inline docker/gcsfuse/Dockerfile:1]: How should this Dockerfile be used? As a base image for components?

File: express/storage_mount.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    mount_args: Optional[List[str]] = None,
+    mount_kwargs: Optional[Dict[str, any]] = None,
+):
+    """
+    Function that mounts a remote bucket to a local directory using FUSE
+    Args:
+        mount_buckets (str): the name of the bucket(s) to be mount
+        mount_dir (str): the local directory where to mount the buckets to
+        mount_command (str): the command used to mount the storage system
+        mount_args (Optional[List[str]]): Additional arguments to pass to the FUSE command.
```
Comment (RobbeSneyders) [inline express/storage_mount.py:56]: Which args and kwargs do we expect the user might want to pass here?

File: express/io.py [hunk 2]
```diff
@@ -31,3 +49,38 @@ def get_file_name(file_uri: str, return_extension=False):
     if return_extension:
         file_name = f"{file_name}{extension}"
     return file_name
+
+
+def create_subprocess_arguments(
+    args: Optional[List[str]] = None, kwargs: Optional[Dict[str, Any]] = None
+) -> List[str]:
```
Comment (RobbeSneyders) [inline express/io.py:56]: I feel like we're taking quite some assumptions here on how arguments will be parsed, which can often differ between programs. If we want the user to be able to pass in arguments, I would let them pass them in already formatted.

File: express/components/common.py [hunk 3]
```diff
@@ -428,15 +485,43 @@ def run(cls) -> DataManifest:
...
-        output_dataset_draft = cls.load(extra_args=json.loads(args.extra_args))
-        metadata = cls._create_metadata(metadata_args=json.loads(args.metadata_args))
-        # Create metadata
+        storage_args = json.loads(args.storage_args)
+        extra_args = json.loads(args.extra_args)
+        metadata_args = json.loads(args.metadata_args)
+
+        # Get specific cloud storage configs
+        cloud_storage = CloudProvider[storage_args["cloud_env"]].value
+        storage_mount_cmd, storage_prefix = (
```
Comment (RobbeSneyders) [inline express/components/common.py:496]: Same here.

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @PhilippeMoussalli 

High level, I think this is a good approach. I left some comments in line.

Some general comments:
- Let's leverage `pathlib` everywhere instead of `os.path` and accept `Path` type where relevant.
- Let's leverage keyword only arguments whenever useful: https://blog.ml6.eu/designing-better-python-interfaces-with-keyword-only-arguments-61888040d971
- Remove typing from the docstrings to have a single source of truth

We can tackle these in a separate PR across the whole codebase as well.

File: express/storage_mount.py (hunk 1 above)
Comment (PhilippeMoussalli) [reply express/storage_mount.py:56]: There are quite a few extra arguments that can be passed to the fuse commands (e.g. passing in keys, enable additional logging, enable debugging). And I think they are all different depending on which cloud you use for fuse. We can maybe restrict the options once we figure out how we want to handle the authentication and enable other ones with extra argument flags that we pass in the function

Comment (RobbeSneyders) [reply express/storage_mount.py]: Question is which ones we want to expose to the user and which one we want to configure.

File: docker/gcsfuse/Dockerfile
Comment (PhilippeMoussalli) [reply docker/gcsfuse/Dockerfile:1]: I think we should have a base image component for Express that contains common system dependencies. In this case, it should be all the Fuse installations for the cloud providers. All the other images will build on top of this image.

Do you agree with this approach?

File: express/components/common.py [hunk 4]
```diff
@@ -72,24 +74,30 @@ def load(
...
+            **kwargs
         )
 
     @staticmethod
     @abstractmethod
     def _load_data_source(
-        data_source: DataSource,
-        index_filter: Optional[IndexT],
-        **kwargs,
+            *,
```
Comment (NielsRogge) [inline express/components/common.py:86]: Same comment here

File: docker/gcsfuse/Dockerfile
Comment (NielsRogge) [reply docker/gcsfuse/Dockerfile:1]: I think that's a good idea

File: express/components/common.py [hunk 5]
```diff
@@ -43,23 +43,25 @@ def extend(self) -> "ExpressDatasetDraft[IndexT, DataT]":
...
+    def load_index(self, mount_dir: str) -> IndexT:
         """
         Loads the index data.
+        Args:
+            mount_dir: the local directory mounted with fuse
         """
 
     def load(
-        self, data_source: str, for_index: Optional[IndexT] = None, **kwargs
+            self, data_source: str, *, for_index: Optional[IndexT] = None, **kwargs
```
[the remark this answers is not in the document]
Comment (RobbeSneyders) [reply express/components/common.py:54]: Why does this make it harder to debug? Keyword-only arguments add more clarity to the code base. I wrote a blog post on this which I linked above:
https://blog.ml6.eu/designing-better-python-interfaces-with-keyword-only-arguments-61888040d971

[the remark this answers is not in the document]
Comment (RobbeSneyders) [reply express/components/common.py:54]: Yes indeed.

**Review by NielsRogge — APPROVED** [review]
PR looks good to me, but maybe it might be easier to integrate after #27 gets merged

**Pull request closed without merging**